### ISSUE TITLE
Epoch scoring can use up a lot of GPU memory

### ISSUE THREAD

**Issue by Mathieu-Prouveur** [skorch-dev/skorch#509]
Hi developpers!

Thanks for the great package.

I recently had a classification multi label setting with lots of targets (~100.000) .
Training was running fine but it crashed during the validation steps and I could see the allocated GPU memory skyrocket during validation steps before it crashed.

The problem came from the `EpochScoring` callback as it stores `y_pred` as returned by the model. I initialized with device='cuda' so as to use my model on GPU, `y_pred` was then stored on GPU. The problem is then that the `EpochScoring` callback stores pointers to these tensors. Thus even though some GPU memory is freed during the validation step, if you have the `y_pred` tensor corresponding to all the validation dataset which is big enough, a lot of GPU memory will be allocated during validation steps and will eventually cause a memory error.

I fixed it by sending `y_pred` to cpu in the `validation_step` but that wouldn't fix the problem if someone used epoch scoring for training steps as well.

I guess it's an edge case as you need to have lots of targets or lots of data to have this issue but I figured it's a good to know

**Comment by BenjaminBossan** [comment]
This a tradeoff between speed and memory. You can turn off caching by setting `use_caching=False` (if you have several `EpochScoring`s, turn it off for all).

**Comment by Mathieu-Prouveur** [comment]
Oh right! Skorch does provide a solution then. But it means we compute again the predictions just to send them on cpu so I think we waste a bit of time there.

**Comment by BenjaminBossan** [comment]
Yeah. There probably isn't a solution that fits each possible use case.

As long as the solution you mentioned initially isn't too hacky, I see no problem with using just that. The idea has always been that skorch should be easy to extend.

**Comment by Mathieu-Prouveur** [comment]
Sure. Thanks for the insights!

**Comment by mshuaibii** [comment]
> Hi developpers!
> 
> Thanks for the great package.
> 
> I recently had a classification multi label setting with lots of targets (~100.000) .
> Training was running fine but it crashed during the validation steps and I could see the allocated GPU memory skyrocket during validation steps before it crashed.
> 
> The problem came from the `EpochScoring` callback as it stores `y_pred` as returned by the model. I initialized with device='cuda' so as to use my model on GPU, `y_pred` was then stored on GPU. The problem is then that the `EpochScoring` callback stores pointers to these tensors. Thus even though some GPU memory is freed during the validation step, if you have the `y_pred` tensor corresponding to all the validation dataset which is big enough, a lot of GPU memory will be allocated during validation steps and will eventually cause a memory error.
> 
> I fixed it by sending `y_pred` to cpu in the `validation_step` but that wouldn't fix the problem if someone used epoch scoring for training steps as well.
> 
> I guess it's an edge case as you need to have lots of targets or lots of data to have this issue but I figured it's a good to know

I'm actually running into this issue in both the train/validation steps. The obvious solution is `use_caching=False`, wondering if there were any changes/workarounds since this original post?

**Comment by BenjaminBossan** [comment]
@mshuaibii No, there hasn't been any addition since the last discussion.

Thank you a lot for posting this solution. This could be useful for others who are running into this problem. Just be sure to remember that `NeuralNetClassifier` and `NeuralNetBinaryClassifier` have a default `EpochScoring` callback (used for accuracy) which needs to be replaced as well, otherwise there is no memory gain.

(NB: The posted solution will avoid leaving the `y_pred`s on GPU memory, which is the desired effect. However, it will create a copy of the tensor on main memory for each callback. So if you have a lot of `MemEffEpochScoring`s and not so much RAM, the solution might exhaust the RAM.)

For a more general solution, I could imagine to provide a parameter to `EpochScoring` that allows users to modify `y_pred` before it is cached. Usage would look like this:

```python
# same as above
def to_cpu(X):
    if isinstance(X, (tuple, list)):
        return type(X)(to_cpu(x) for x in X)
    return X.detach().to("cpu")

net = NeuralNet(
    ...,
    callbacks=[EpochScoring(..., pre_caching_transform=to_cpu)],
)
```

Probably there is a better name than `pre_caching_transform`. What do you think?

Ping @ottonemo @thomasjpfan
